File: ocsgw/src/main/java/org/ostelco/ocsgw/converter/ProtobufToDiameterConverter.java
package org.ostelco.ocsgw.converter;

import com.google.protobuf.ByteString;
import org.ostelco.diameter.CreditControlContext;
import org.ostelco.diameter.model.*;
import org.ostelco.ocs.api.CreditControlRequestInfo;
import org.ostelco.ocs.api.CreditControlRequestType;
import org.ostelco.ocs.api.PsInformation;
import org.ostelco.ocs.api.ServiceInfo;
import org.ostelco.ocsgw.datasource.protobuf.GrpcDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.util.Collections;

import static org.ostelco.diameter.model.RequestType.*;

public class ProtobufToDiameterConverter {

    private static final Logger LOG = LoggerFactory.getLogger(GrpcDataSource.class);

    /**
     * Convert MultipleServiceCreditControl in gRPC format to diameter format
     */
    public static MultipleServiceCreditControl convertMSCC(org.ostelco.ocs.api.MultipleServiceCreditControl msccGRPC) {
        return new MultipleServiceCreditControl(
                msccGRPC.getRatingGroup(),
                (int) msccGRPC.getServiceIdentifier(),
                Collections.singletonList(new ServiceUnit()),
                Collections.singletonList(new ServiceUnit()),
                new ServiceUnit(msccGRPC.getGranted().getTotalOctets(), 0, 0),
                msccGRPC.getValidityTime(),
                msccGRPC.getQuotaHoldingTime(),
                msccGRPC.getVolumeQuotaThreshold(),
                convertFinalUnitIndication(msccGRPC.getFinalUnitIndication()),
                convertResultCode(msccGRPC.getResultCode()));
    }

    /**
     * Convert Diameter request type to gRPC
     */
    public static CreditControlRequestType getRequestType(CreditControlContext context) {
        switch (context.getOriginalCreditControlRequest().getRequestTypeAVPValue()) {
            case INITIAL_REQUEST:
                return CreditControlRequestType.INITIAL_REQUEST;
            case UPDATE_REQUEST:
                return CreditControlRequestType.UPDATE_REQUEST;
            case TERMINATION_REQUEST:
                return CreditControlRequestType.TERMINATION_REQUEST;
            case EVENT_REQUEST:
                return CreditControlRequestType.EVENT_REQUEST;
            default:
                LOG.warn("Unknown request type");
                return CreditControlRequestType.NONE;
        }
    }

    private static FinalUnitIndication convertFinalUnitIndication(org.ostelco.ocs.api.FinalUnitIndication fuiGrpc) {
        if (!fuiGrpc.getIsSet()) {
            return null;
        }
        return new FinalUnitIndication(
                FinalUnitAction.values()[fuiGrpc.getFinalUnitAction().getNumber()],
                fuiGrpc.getRestrictionFilterRuleList(),
                fuiGrpc.getFilterIdList(),
                new RedirectServer(
                        RedirectAddressType.values()[fuiGrpc.getRedirectServer().getRedirectAddressType().getNumber()],
                        fuiGrpc.getRedirectServer().getRedirectServerAddress()
                )
        );
    }

    // We match the error codes on names in protobuf and internal model
    public static ResultCode convertResultCode(org.ostelco.ocs.api.ResultCode resultCode) {
        return ResultCode.valueOf(resultCode.name());
    }

    public static CreditControlRequestInfo convertRequestToProtobuf(final CreditControlContext context, @Nullable final String topicId) {

        try {
            CreditControlRequestInfo.Builder builder = CreditControlRequestInfo
                    .newBuilder()
                    .setType(getRequestType(context));

            if (topicId != null) {
                builder.setTopicId(topicId);
            }

            builder.setRequestNumber(context.getCreditControlRequest().getCcRequestNumber().getInteger32());

            addMultipleServiceCreditControl(context, builder);

            builder.setRequestId(context.getSessionId())
                    .setMsisdn(context.getCreditControlRequest().getMsisdn())
                    .setImsi(context.getCreditControlRequest().getImsi());

            addPsInformation(context, builder);

            return builder.build();

        } catch (Exception e) {
            LOG.error("Failed to create CreditControlRequestInfo [{}] [{}]", context.getCreditControlRequest().getMsisdn(), context.getSessionId(), e);
        }
        return null;
    }

    private static void addMultipleServiceCreditControl(final CreditControlContext context, CreditControlRequestInfo.Builder builder) {
        for (MultipleServiceCreditControl mscc : context.getCreditControlRequest().getMultipleServiceCreditControls()) {

            org.ostelco.ocs.api.MultipleServiceCreditControl.Builder protoMscc = org.ostelco.ocs.api.MultipleServiceCreditControl.newBuilder();

            if (!mscc.getRequested().isEmpty()) {

                ServiceUnit requested = mscc.getRequested().get(0);

                protoMscc.setRequested(org.ostelco.ocs.api.ServiceUnit.newBuilder()
                        .setTotalOctets(requested.getTotal()) // fails at 55904
                        .setInputOctets(0L)
                        .setOutputOctets(0L));
            }

            for (ServiceUnit used : mscc.getUsed()) {

                // We do not track CC-Service-Specific-Units or CC-Time
                if (used.getTotal() > 0) {
                    protoMscc.setUsed(org.ostelco.ocs.api.ServiceUnit.newBuilder()
                            .setInputOctets(used.getInput())
                            .setOutputOctets(used.getOutput())
                            .setTotalOctets(used.getTotal()));
                }
            }

            protoMscc.setRatingGroup(mscc.getRatingGroup());
            protoMscc.setServiceIdentifier(mscc.getServiceIdentifier());

            if (mscc.getReportingReason() != null) {
                protoMscc.setReportingReasonValue(mscc.getReportingReason().ordinal());
            } else {
                protoMscc.setReportingReasonValue(org.ostelco.ocs.api.ReportingReason.UNRECOGNIZED.ordinal());
            }
            builder.addMscc(protoMscc);
        }
    }

    private static void addPsInformation(final CreditControlContext context, CreditControlRequestInfo.Builder builder) {
        if (!context.getCreditControlRequest().getServiceInformation().isEmpty()) {
            final org.ostelco.diameter.model.PsInformation psInformation
                    = context.getCreditControlRequest().getServiceInformation().get(0).getPsInformation().get(0);

            if (psInformation != null) {
                PsInformation.Builder psInformationBuilder  = org.ostelco.ocs.api.PsInformation.newBuilder();
                if (psInformation.getCalledStationId() != null) {
                    psInformationBuilder.setCalledStationId(psInformation.getCalledStationId());
                }
                if (psInformation.getSgsnMccMnc() != null) {
                    psInformationBuilder.setSgsnMccMnc(psInformation.getSgsnMccMnc());
                }
                if (psInformation.getImsiMccMnc() != null) {
                    psInformationBuilder.setImsiMccMnc(psInformation.getImsiMccMnc());
                }
                if (psInformation.getUserLocationInfo() != null) {
                    psInformationBuilder.setUserLocationInfo(ByteString.copyFrom(psInformation.getUserLocationInfo()));
                }
                if (psInformation.getPdpAddress() != null) {
                    psInformationBuilder.setPdpAddress(psInformation.getPdpAddress().getHostAddress());
                }
                builder.setServiceInformation(ServiceInfo.newBuilder().setPsInformation(psInformationBuilder));
            }
        }
    }
}

File: ocsgw/src/main/kotlin/org/ostelco/ocsgw/converter/ProtobufToDiameterConverter.kt
package org.ostelco.ocsgw.converter;

import com.google.protobuf.ByteString
import org.ostelco.diameter.CreditControlContext
import org.ostelco.diameter.getLogger
import org.ostelco.diameter.model.FinalUnitAction
import org.ostelco.diameter.model.FinalUnitIndication
import org.ostelco.diameter.model.MultipleServiceCreditControl
import org.ostelco.diameter.model.RedirectAddressType
import org.ostelco.diameter.model.RedirectServer
import org.ostelco.diameter.model.ResultCode
import org.ostelco.diameter.model.ServiceUnit
import org.ostelco.diameter.model.RequestType
import org.ostelco.ocs.api.*
import org.ostelco.ocs.api.PsInformation
import org.ostelco.ocs.api.ReportingReason


class ProtobufToDiameterConverter {

    companion object {

        private val logger by getLogger()

        fun convertRequestToProtobuf(context: CreditControlContext, topicId: String?): CreditControlRequestInfo? {
            try {
                val builder = CreditControlRequestInfo
                        .newBuilder()
                        .setType(getRequestType(context))
                if (topicId != null) {
                    builder.topicId = topicId
                }
                builder.requestNumber = context.creditControlRequest.ccRequestNumber!!.integer32
                addMultipleServiceCreditControl(context, builder)
                builder.setRequestId(context.sessionId)
                        .setMsisdn(context.creditControlRequest.msisdn)
                        .setImsi(context.creditControlRequest.imsi)
                addPsInformation(context, builder)
                return builder.build()
            } catch (e: Exception) {
                logger.error("Failed to create CreditControlRequestInfo [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
            }
            return null
        }

        /**
         * Convert MultipleServiceCreditControl in protobuf format to diameter format
         */
        fun convertMSCC(msccGRPC: org.ostelco.ocs.api.MultipleServiceCreditControl): MultipleServiceCreditControl {
            return MultipleServiceCreditControl(
                    msccGRPC.ratingGroup,
                    msccGRPC.serviceIdentifier,
                    listOf(ServiceUnit()),
                    listOf(ServiceUnit()),
                    ServiceUnit(msccGRPC.granted.totalOctets, 0, 0),
                    msccGRPC.validityTime,
                    msccGRPC.quotaHoldingTime,
                    msccGRPC.volumeQuotaThreshold,
                    convertFinalUnitIndication(msccGRPC.finalUnitIndication),
                    convertResultCode(msccGRPC.resultCode))
        }

        // We match the error codes on names in protobuf and internal model
        fun convertResultCode(resultCode: org.ostelco.ocs.api.ResultCode): ResultCode {
            try {
                return ResultCode.valueOf(resultCode.name)
            } catch (e: IllegalArgumentException) {
                logger.warn("Unknown ResultCode $resultCode")
            }
            return ResultCode.DIAMETER_UNABLE_TO_COMPLY
        }

        /**
         * Convert Diameter request type to protobuf
         */
        fun getRequestType(context: CreditControlContext): CreditControlRequestType {
            return when (context.originalCreditControlRequest.requestTypeAVPValue) {
                RequestType.INITIAL_REQUEST -> CreditControlRequestType.INITIAL_REQUEST
                RequestType.UPDATE_REQUEST -> CreditControlRequestType.UPDATE_REQUEST
                RequestType.TERMINATION_REQUEST -> CreditControlRequestType.TERMINATION_REQUEST
                RequestType.EVENT_REQUEST -> CreditControlRequestType.EVENT_REQUEST
                else -> {
                    logger.warn("Unknown request type")
                    CreditControlRequestType.NONE
                }
            }
        }

        private fun addPsInformation(context: CreditControlContext, builder: CreditControlRequestInfo.Builder) {
            if (!context.creditControlRequest.serviceInformation.isEmpty()) {
                val psInformation = context.creditControlRequest.serviceInformation[0].psInformation[0]
                val psInformationBuilder = PsInformation.newBuilder()
                if (psInformation.calledStationId != null) {
                    psInformationBuilder.calledStationId = psInformation.calledStationId
                }
                if (psInformation.sgsnMccMnc != null) {
                    psInformationBuilder.sgsnMccMnc = psInformation.sgsnMccMnc
                }
                if (psInformation.imsiMccMnc != null) {
                    psInformationBuilder.imsiMccMnc = psInformation.imsiMccMnc
                }

                psInformation.userLocationInfo?.let { psInformationBuilder.userLocationInfo = ByteString.copyFrom(it) }

                psInformation.pdpType?.let { psInformationBuilder.pdpType = PdpType.forNumber(it) }

                psInformation.pdpAddress?.let { psInformationBuilder.pdpAddress = ByteString.copyFrom(it.address) }

                builder.setServiceInformation(ServiceInfo.newBuilder().setPsInformation(psInformationBuilder))
            }
        }

        private fun addMultipleServiceCreditControl(context: CreditControlContext, builder: CreditControlRequestInfo.Builder) {
            for (mscc in context.creditControlRequest.multipleServiceCreditControls) {
                val msccBuilder = org.ostelco.ocs.api.MultipleServiceCreditControl.newBuilder()
                if (!mscc.requested.isEmpty()) {
                    val requested = mscc.requested[0]
                    msccBuilder.setRequested(org.ostelco.ocs.api.ServiceUnit.newBuilder()
                            .setTotalOctets(requested.total)
                            .setInputOctets(0L)
                            .setOutputOctets(0L))
                }
                for (used in mscc.used) { // We do not track CC-Service-Specific-Units or CC-Time
                    if (used.total > 0) {
                        msccBuilder.setUsed(org.ostelco.ocs.api.ServiceUnit.newBuilder()
                                .setInputOctets(used.input)
                                .setOutputOctets(used.output)
                                .setTotalOctets(used.total))
                    }
                }
                msccBuilder.ratingGroup = mscc.ratingGroup
                msccBuilder.serviceIdentifier = mscc.serviceIdentifier

                val reportingReason = mscc.reportingReason
                if (reportingReason != null) {
                    msccBuilder.reportingReasonValue = reportingReason.ordinal
                } else {
                    msccBuilder.reportingReasonValue = ReportingReason.UNRECOGNIZED.ordinal
                }
                builder.addMscc(msccBuilder)
            }
        }

        private fun convertFinalUnitIndication(fuiGrpc: org.ostelco.ocs.api.FinalUnitIndication): FinalUnitIndication? {
            return if (!fuiGrpc.isSet) {
                null
            } else FinalUnitIndication(
                    FinalUnitAction.values()[fuiGrpc.finalUnitAction.number],
                    fuiGrpc.restrictionFilterRuleList,
                    fuiGrpc.filterIdList,
                    RedirectServer(
                            RedirectAddressType.values()[fuiGrpc.redirectServer.redirectAddressType.number],
                            fuiGrpc.redirectServer.redirectServerAddress
                    )
            )
        }
    }
}
